P2798-fix-default-constructed-mappings/P2798.md

```diff
@@ -0,0 +1,56 @@
+---
+title: "Fix layout mappings all static extent default constructor"
+document: P2798R0
+date: today
+audience: LWG
+author:
+  - name: Christian Trott 
+    email: <crtrott@sandia.gov>
+  - name: Damien Lebrun-Grandie 
+    email: <lebrungrandt@ornl.gov>
+  - name: Mark Hoemmen 
+    email: <mhoemmen@nvidia.com>
+toc: true
+---
+
+
+# Revision History
+
+## Initial Version 2023-02 Mailing
+
+# Rationale
+
+When default constructing a `layout_left`, `layout_right` or `layout_stride` mapping with all static extents,
+there is currently no precondition check guarding against overflow of the `index_type`.
+
+A simple example is constructing the following `layout_left` mapping:
+
+```c++
+constexpr size_t N = 4'000'000;
+layout_left::mapping<extents<int, N, N>> map;
+```
+
+Each individual extent fits into `int` but the theoretical required span size does not.
+
+The non-default constructors already deal with this issue via a preconditions such as:
+
+*Preconditions:* The size of the multidimensional index space `e` is representable as a value of type `index_type` ([basic.fundamental]).
+
+For cases where at least one of the extents is dynamic, the default constructor is also fine, since it will result in an empty index space. 
```

Comment:
Is it, though?  What about the following `extents`?

```c++
extents<int, 4'000'000, dynamic_extent, 4'000'000>;
```

Product of the extents is zero mathematically, but will overflow.  Do we need to worry about this case?

I think we can fix this by saying that the forward product of the extents is representable as `index_type`, regardless of whether the extents are dynamic or static.  This would generalize Tomasz's suggestion that we treat `rank_dynamic() == 0` and `rank_dynamic() != 0` in the same way.

Reply:
we can't do that. Because the extent for dynamic thing is bad. I.e. its like max size_t value -1 or so. Also This doesn't say how to compute the size of multi-dimensional index space. Its more like the pure math expression.